cmd/kube-proxy/app/server_test.go

```diff
@@ -279,35 +281,130 @@ nodePortAddresses:
+	duplicateKeyYAMLTemplate := `bindAddress: 0.0.0.0
+bindAddress: 1.2.3.4
+clientConnection:
+  acceptContentTypes: "abc"
+  burst: 100
+  contentType: content-type
+  kubeconfig: "/path/to/kubeconfig"
+  qps: 7
+clusterCIDR: "1.2.3.0/24"
+configSyncPeriod: 15s
+conntrack:
+  maxPerCore: 2
+  min: 1
+  tcpCloseWaitTimeout: 10s
+  tcpEstablishedTimeout: 20s
+healthzBindAddress: "1.2.3.4:12345"
+hostnameOverride: "foo"
+iptables:
+  masqueradeAll: true
+  masqueradeBit: 17
+  minSyncPeriod: 10s
+  syncPeriod: 60s
+ipvs:
+  minSyncPeriod: 10s
+  syncPeriod: 60s
+  excludeCIDRs:
+    - "10.20.30.40/16"
+    - "fd00:1::0/64"
+kind: KubeProxyConfiguration
+metricsBindAddress: "1.2.3.4:12346"
+mode: "loopback"
+oomScoreAdj: 17
+portRange: "2-7"
+udpIdleTimeout: 123ms
+nodePortAddresses:
+  - "10.20.30.40/16"
+  - "fd00:1::0/64"`
+
+	unknownFieldYAMLTemplate := `bindAddress: 0.0.0.0
+foo: bar
+clientConnection:
+  acceptContentTypes: "abc"
+  burst: 100
+  contentType: content-type
+  kubeconfig: "/path/to/kubeconfig"
+  qps: 7
+clusterCIDR: "1.2.3.0/24"
+configSyncPeriod: 15s
+conntrack:
+  maxPerCore: 2
+  min: 1
+  tcpCloseWaitTimeout: 10s
+  tcpEstablishedTimeout: 20s
+healthzBindAddress: "1.2.3.4:12345"
+hostnameOverride: "foo"
+iptables:
+  masqueradeAll: true
+  masqueradeBit: 17
+  minSyncPeriod: 10s
+  syncPeriod: 60s
+ipvs:
+  minSyncPeriod: 10s
+  syncPeriod: 60s
+  excludeCIDRs:
+    - "10.20.30.40/16"
+    - "fd00:1::0/64"
+kind: KubeProxyConfiguration
+metricsBindAddress: "1.2.3.4:12346"
+mode: "loopback"
+oomScoreAdj: 17
+portRange: "2-7"
+udpIdleTimeout: 123ms
+nodePortAddresses:
+  - "10.20.30.40/16"
+  - "fd00:1::0/64"`
+
 	testCases := []struct {
 		name   string
 		config string
-		expErr string
+		errFn  func(err error) bool
 	}{
 		{
 			name:   "Decode error test",
 			config: "Twas bryllyg, and ye slythy toves",
-			expErr: "could not find expected ':'",
+			errFn:  func(err error) bool { return err != nil },
 		},
 		{
 			name:   "Bad config type test",
 			config: "kind: KubeSchedulerConfiguration",
-			expErr: "no kind",
+			errFn:  func(err error) bool { return err != nil },
 		},
 		{
 			name:   "Missing quotes around :: bindAddress",
 			config: "bindAddress: ::",
-			expErr: "mapping values are not allowed in this context",
+			errFn:  func(err error) bool { return err != nil },
+		},
+		{
+			name:   "Duplicate fields",
+			config: duplicateKeyYAMLTemplate,
+			errFn: func(err error) bool {
+				return err != nil && kuberuntime.IsStrictDecodingError(err)
+			},
+		},
+		{
+			name:   "Unknown field",
+			config: unknownFieldYAMLTemplate,
+			errFn: func(err error) bool {
+				return err != nil && kuberuntime.IsStrictDecodingError(err)
+			},
 		},
 	}
+
 	version := "apiVersion: kubeproxy.config.k8s.io/v1alpha1"
 	for _, tc := range testCases {
-		options := NewOptions()
-		config := fmt.Sprintf("%s\n%s", version, tc.config)
-		_, err := options.loadConfig([]byte(config))
-		if assert.Error(t, err, tc.name) {
```

Comment:
Nice addition of `t.Run` :+1:

We should keep this:
```go
if assert.Error(t, err, tc.name) {
```
This whole test is about failures, so we assert that one should exist.

We can keep the old test behavior for string matching the error text.
To gate it, just check for the zero value `""`.
The gate is not strictly necessary since Contains will always pass with an empty string, but it's more clear.

We should also check that `errFn != nil` before running it, otherwise go will
`panic: runtime error: invalid memory address or nil pointer dereference`.
Gating this has the added benefit of making this `errFn` optional as well.

Finally, we should use the assert library to maintain style.

ex:
```go
t.Run(tc.name, func(t *testing.T) {
	options := NewOptions()
	config := fmt.Sprintf("%s\n%s", version, tc.config)
	_, err := options.loadConfig([]byte(config))
	if assert.Error(t, err, tc.name) {
		if tc.expErr != "" {
			assert.Contains(t, err.Error(), tc.expErr, tc.name)
		}
		if tc.errFn != nil {
			assert.True(t, tc.errFn(err), tc.name)
		}
	}
})
```

Reply:
`errFn` may be better named as `checkFn`.
The verb/purpose is not "to error" but rather "to check and return a bool".

Reply:
I like it much better! I adjusted it accordingly